test/e2e/framework/service/jig.go

```diff
@@ -390,10 +392,57 @@ func (j *TestJig) waitForAvailableEndpoint(timeout time.Duration) error {
 
 	go controller.Run(stopCh)
 
-	err := wait.Poll(1*time.Second, timeout, func() (bool, error) {
-		return endpointAvailable, nil
-	})
-	if err != nil {
+	// If EndpointSlice API is enabled, then validate if appropriate EndpointSlice objects were also create/updated/deleted.
+	if _, err := j.Client.Discovery().ServerResourcesForGroupVersion(discoveryv1beta1.SchemeGroupVersion.String()); err == nil {
```

Comment:
maybe:

```
if _, err := j.Client.Discovery().ServerResourcesForGroupVersion(discoveryv1beta1.SchemeGroupVersion.String()); err != nil {
  // EndpointSlice API is disabled.
  return nil
}

// everything else unintended thus easier to parse.

Reply:
So this is unfortunately messy to check... The API being enabled does not actually mean that the controller is running, just that it can be running. Checking to see if an EndpointSlice with the name `kubernetes` existed would tell you that the `EndpointSlice` feature gate was enabled on API Server, but that's about all we've got. So with that said, the check you already have is probably as much as we can do here, but it does mean we may end up waiting indefinitely for EndpointSlices in some cases.

Reply:
@wojtek-t I'm creating a new variable inside the `if` block ... and the check is different using slices than not

@robscott we had to add this check https://github.com/kubernetes/kubernetes/pull/93597 because it was failing in a job that only enable GA features (thanks to Jordan for adding it). I think that for testing should be fair assume API enable = ESController Running,... when can that not happen?

Reply:
@aojea I think that is pretty unlikely in the future, but it has happened in the past. In 1.17 the API was enabled by default but the controller and feature gate were not. It wasn't until 1.18 that both were enabled by default. I don't _think_ there are any test environments that would have this kind of weird configuration at this point.

Reply:
Yeah - I agree that enablement of the API is not the same as controller, but I think that for the purposes of e2e tests I think it's actually a good enough assumption.


> @wojtek-t I'm creating a new variable inside the if block ... and the check is different using slices than not

@aojea  I don't get it. I'm not asking to check any of the implemention. All I'm saying is:
- instead o having
```
if <condition> {
  // contents
}
```
have
```
if !<condition> {
  return nil
}
<contents>
```

Reply:
Actually - I see what you're saying now. Sorry.

So let me suggest slightly different approach
```
endpointAvailable := false
endpointSlicesAvailable := false

startEndpointInformer(stopCh)
if <api enabled> {
  startEndpointSliceInformer(stopCh)
} else {
  endpointSliceAvailable = true
}

err := wait.Poll(1*time.Second, timeout, func() (bool, error) {
  return endpointAvailable && endpointSliceAvailable, nil
})
...
```

Reply:
done